src/renderers/webgl/WebGLPrograms.js

```diff
@@ -296,8 +299,8 @@ function WebGLPrograms( renderer, cubemaps, cubeuvmaps, extensions, capabilities
 
 		} else {
 
-			array.push( hashString( parameters.fragmentShader ) );
-			array.push( hashString( parameters.vertexShader ) );
+			array.push( parameters.fragmentShaderID );
+			array.push( parameters.vertexShaderID );
```

Comment:
The idea of using ids instead of hash strings is great! However, I vote for an implementation that does not affect material classes and is instead hidden in the renderer.

As an alternative, I've implemented an additional shader cache for custom materials in `WebGLPrograms`. I guess this would lead to the same performance improvement since `hashString()` is not used anymore. 

https://github.com/Mugen87/three.js/commit/dae408807836c67b06065d9432c57ee2440c3f4a

Reply:
I like your approach better although I wouldn't map shaders to the material. Since that would mean that the same shaders in different materials would have different IDs which would create different programs. This is noticeable in my example because it takes a lot longer to load and It creates 2500 programs in the meantime.

Example of your implementation  => [link](https://raw.githack.com/gero3/three.js/MugenWebglShaderCache/examples/webgl_performance_shader.html)
![image](https://user-images.githubusercontent.com/155535/145657671-e9d673d4-b060-4ad4-a1f8-81dd8bc0ea82.png)

I think we can write a decently fast version of WebglShaderCache that doesn't take into account the material for the shaderID.

Should I close this PR and rewrite your implementation with the dependence on the material?

Reply:
I just tested the performance on my galaxy S10e which had 15fps for this PR and 12 fps for @Mugen87 implementation of webglshadercache. The examples on my other systems get 60fps so I can't really see the difference anymore on those systems.

Reply:
>  like your approach better although I wouldn't map shaders to the material. 

That is a good point!

I've updated my branch with a small extension similar to what we have for `WebGPURenderer`. The idea is to have an additional class `WebGLShaderStage` to identify shader stages and to avoid a direct shader->material mapping.

https://github.com/Mugen87/three.js/commit/390f7efe1d3b806fe6e21c5e7e276f3148acba29

Reply:
Seems good to me. 
[link](https://raw.githack.com/gero3/three.js/mugenshaderstagecacheimplementation/examples/webgl_performance_shader.html)

Reply:
Some more performance data with my iMac:

Dev: 37 FPS
This PR: 48 FPS
Using `WebGLShaderCache`: 48 FPS

Reply:
Dev 40FPS
PR 59FPS
ShaderCache 58FPS

I like Mugen's approach of mimic-ing WebGPU's shader cache stage, should allow for a more natural extension and safer transition down the road.

Reply:
@sciecode Indeed! The more we can mimic WebGPU the better.